README.md
<img src="./images/IFX_LOGO_600.gif" align="right" width="150"/>

# ADC Calibration

**This code example demonstrates the ADC calibration algorithm.**  

## Device

The device used in this code example (CE) is:
- [TRAVEO™ T2G CYT4BF Series](https://www.infineon.com/cms/en/product/microcontroller/32-bit-traveo-t2g-arm-cortex-microcontroller/32-bit-traveo-t2g-arm-cortex-for-body/traveo-t2g-cyt4bf-series/)

## Board

The board used for testing is:
- TRAVEO™ T2G evaluation kit ([KIT_T2G-B-H_LITE](https://www.infineon.com/cms/en/product/evaluation-boards/kit_t2g-b-h_lite/))

## Scope of work

In this example, the ADC calibration algorithm is executed to find an optimal offset and gain compensation for the ADC. It is also providing the options for decalibration by setting different offset and gain values (boundary values) which result in a decalibrated analog-to-digital transfer curve. A single ADC conversion is showing the voltage level of a potentiometer output while the calibration is done on a different logical ADC channel. 

The ADC calibration is recommended to be executed regularly to compensate for environmental effects (e.g. temperature) on the analog-to-digital transfer curve.

## Introduction  

**ADC calibration**  

TRAVEO™ T2G platform supports the following ADC calibration features:
- Ability to configure offset and gain to make the actual ADC transfer curve come closer to the ideal transfer curve.
- The compensation can be adjusted by an 8-bit (-128 to 127) analog offset and a 5-bit (-15 to 15) analog gain.
- The offset and gain adjustment step is a quarter of 1 LSB.
- Two different register sets for calibration are supported, the regular calibration and the alternate calibration. An update can be triggered to copy the data from the alternate set to the regular set.

More details can be found in [Technical Reference Manual (TRM)](https://www.infineon.com/dgdl/?fileId=5546d4627600a6bc017600bfae720007), [Registers TRM](https://www.infineon.com/dgdl/?fileId=5546d4627600a6bc017600be2aef0004) and [Data Sheet](https://www.infineon.com/dgdl/?fileId=5546d46275b79adb0175dc8387f93228).

## Hardware setup

This CE has been developed for:
- TRAVEO™ T2G evaluation kit lite ([KIT_T2G-B-H_LITE](https://www.infineon.com/cms/en/product/evaluation-boards/kit_t2g-b-h_lite/))<BR>
<img src="./images/KIT_T2G-B-H_LITE.gif"/><BR>  
No changes are required from the board's default settings.

The on-board potentiometer *VR1* can be used to define a voltage level connected to an ADC input channel. It can also be used to prove the success of an ADC calibration.

Please refer to the board schematic for more details on how the potentiometer is connected to the ADC.

## Implementation

This design consists of providing different options: Triggering a single ADC conversion, execution of ADC calibration and changing offset and gain to decalibrate the analog-to-digital transfer curve. The terminal program displays a menu and the user can choose an option by entering the corresponding key.

The ADC block is configured for two logical channels. One is used for ADC conversion of the potentiometer output voltage level. The second channel is used for the ADC calibration where internal reference voltage levels are set as an ADC input.

**STDOUT/STDIN setting**

Initialization of the GPIO for UART is done in the <a href="https://infineon.github.io/retarget-io/html/group__group__board__libs.html#gaddff65f18135a8491811ee3886e69707"><i>cy_retarget_io_init()</i></a> function.
- Initializes the pin specified by *CYBSP_DEBUG_UART_TX* as UART TX and the pin specified by *CYBSP_DEBUG_UART_RX* as UART RX (these pins are connected to KitProg3 COM port)
- The serial port parameters are 8N1 and 115200 baud

**ADC initialization**

API calls for the ADC initialization are done in the *main()* function based on the application note [AN219755](https://www.infineon.com/dgdl/?fileId=8ac78c8c7cdc391c017d0d3aaebf676a).
- Set ePASS MMIO reference buffer mode for bandgap voltage via <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#gab2725bd7acf6658c561a1a4614c12c5d"><i>Cy_SAR2_SetReferenceBufferMode()</i></a>
- Enable the Diagnostic Reference via <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#gaf5f91292a3e1841b4b86f9e6e6788853"><i>Cy_SAR2_Diag_Enable()</i></a>
- For initialization of the ADC, <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#gad6ddc69b82b06fda5952c5da00de35ed"><i>Cy_SAR2_Init()</i></a> is called with appropriate arguments by using macros and configuration structure <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/structcy__stc__sar2__config__t.html"><i>cy_stc_sar2_config_t</i></a> which are auto-generated by the Device Configurator and can be found in *cycfg_peripherals.h*.

  *Figure 1. ADC configuration*<BR><img src="./images/ADC_configuration.png" width="1200"/>
- The last step is to enable the ADC by <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#gae928dc6cfa80e6ee015be6aac6fc759e"><i>Cy_SAR2_Enable()</i></a>

**ADC calibration menu**

The menu is visible in a terminal and printed in the *printMenu()* function.

**Code example main loop**

The endless loop of the code example in *main()* always checks whether a character has been received from the user via UART (terminal) using <a href="https://infineon.github.io/mtb-hal-cat1/html/group__group__hal__uart.html#ga89108b2d339dc9863ec660588e3a4a12"><i>cyhal_uart_getc()</i></a>. The user interface supports different characters:
- *0* triggers a single ADC conversion of the potentiometer voltage level
- *1* executes the ADC calibration algorithm
- *2* de-calibrates the ADC by setting the offset to 127
- *3* de-calibrates the ADC by setting the offset to -128
- *4* de-calibrates the ADC by setting the gain to 15
- *5* de-calibrates the ADC by setting the gain to -15
- Any other key displays the menu

**ADC single conversion**

A single ADC conversion is started in the *getAdcValue()* function which returns the result of the related conversation of the chosen logical ADC channel. It is called whenever the ADC conversion is triggered by the user (press "0" key) or during the ADC calibration functions *calibrateOffset()* and *calibrateGain()*. It triggers a conversion by calling <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga07a7023e4f6db655204d25a21b036651"><i>Cy_SAR2_Channel_SoftwareTrigger()</i></a>. Then it waits for the successful completion of ADC conversion by evaluation of <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#gae07d8e288f6863cef7e8fa37fa2c0f55"><i>Cy_SAR2_Channel_GetInterruptStatus()</i></a>. Since the ADC conversion has been completed the interrupt is cleared by calling <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga3038fbd14b4fef98a91a8713c559472d"><i>Cy_SAR2_Channel_ClearInterrupt()</i></a>. The conversion result is read by calling <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga6cf749ec06d3ab25ce139975f7baafcf"><i>Cy_SAR2_Channel_GetResult()</i></a>.

**ADC calibration algorithm**

The calibration is done in two subsequent steps.
1. ADC offset calibration is executed by calling the function *calibrateOffset()*. It returns "true" when the calibration was successfully performed, and a "false" in case of an error. Associated outputs on the terminal are created (optimal offset or error report). 
The ADC offset calibration sequence is as follows:
    - Set *VREF_L* as source for the channel by changing the ".pinAddress" attribute in <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/structcy__stc__sar2__channel__config__t.html"><i>cy_stc_sar2_channel_config_t</i></a> to *CY_SAR2_PIN_ADDRESS_VREF_L* and call the init function <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga20289a8407d268ade39d00c18a7c14b8"><i>Cy_SAR2_Channel_Init()</i></a>. This connects an internal reference low level (0V) to the ADC input.
    - Set the gain correction setting to "0"
    - Change the ADC offset compensation in a loop from *OFFSET_MAX* (127) to *OFFSET_MIN* (-128) by writing to the alternate calibration register using <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga00e281a1954cef9b30234cf219ea8a69"><i>Cy_SAR2_SetAltAnalogCalibrationValue()</i></a>. 
    - The ideal lower offset *offsetLow* is found when the ADC conversion results in "0"
    - Set *VREF_H* as source for the channel by changing the ".pinAddress" attribute in <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/structcy__stc__sar2__channel__config__t.html"><i>cy_stc_sar2_channel_config_t</i></a> to *CY_SAR2_PIN_ADDRESS_VREF_H* and call the init function <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga20289a8407d268ade39d00c18a7c14b8"><i>Cy_SAR2_Channel_Init()</i></a>. This connects an internal reference high-level (5V) to the ADC input.
    - Change the ADC offset compensation in a loop from *OFFSET_MIN* (-128) to *OFFSET_MAX* (127) by writing to the alternate calibration register using <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga00e281a1954cef9b30234cf219ea8a69"><i>Cy_SAR2_SetAltAnalogCalibrationValue()</i></a>. 
    - The ideal upper offset *offsetHigh* is found when the ADC conversion results in "4095"
    - In the final step, the optimal offset is calculated by the arithmetic mean value between *offsetHigh* and *offsetLow* and some fixed offset<BR><BR>
    *Figure 2. Relationship between OFST and offset shift direction*<BR><img src="./images/offset_compensation.jpg" width="800"/> <BR><BR>
   
2. ADC gain calibration is executed by calling the function *calibrateGain()*. It returns "true" when the calibration was successfully performed, and a "false" in case of an error. Associated outputs on the terminal are created (optimal gain or error report).
The ADC gain calibration sequence is as follows:
    - Set *VREF_L* as source for the channel by changing the ".pinAddress" attribute in <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/structcy__stc__sar2__channel__config__t.html"><i>cy_stc_sar2_channel_config_t</i></a> to *CY_SAR2_PIN_ADDRESS_VREF_L* and call the init function <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga20289a8407d268ade39d00c18a7c14b8"><i>Cy_SAR2_Channel_Init()</i></a>. This connects an internal reference low level (0V) to the ADC input.
    - Change the ADC gain compensation in a loop from *GAIN_MAX* (15) to *GAIN_MIN* (-15) by writing to the alternate calibration register using <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga00e281a1954cef9b30234cf219ea8a69"><i>Cy_SAR2_SetAltAnalogCalibrationValue()</i></a>. 
    - The ideal gain is found when the ADC conversion results in "0"<BR>    
    *Figure 3. Relationship between GAIN and gain shift direction*<BR><img src="./images/gain_compensation.jpg" width="800"/><BR><BR>
    
The ADC is supporting two sets of ADC calibration registers. One regular register is used for the ADC conversions in the user's regular software flow and the alternate register set for ADC calibration in the background. After the ADC calibration was successfully performed, the data in the alternate register set can be copied into the regular calibration set by calling <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga6a00e61ed457bdd7c579a8dfff2a1835"><i>Cy_SAR2_TriggerCalibrationUpdate()</i></a>.

**ADC decalibration**

To prove the behavior of ADC, the offset and gain can be decalibrated by writing some boundary values to the regular calibration register set via <a href="https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__sar2__functions.html#ga2a2919b5813338f4c40a74246cef135e"><i>Cy_SAR2_SetAnalogCalibrationValue()</i></a>.
After changing the offset or gain, the analog-to-digital transfer curve should be decalibrated and showing irregular ADC conversion data.

## Run and Test

For this example, a terminal emulator is required for displaying outputs and to change settings during runtime. Install a terminal emulator if you do not have one. Instructions in this document use [Tera Term](https://ttssh2.osdn.jp/index.html.en).

After code compilation, perform the following steps for flashing the device:
1. Connect the board to your PC using the provided USB cable through the KitProg3 USB connector.
2. Open a terminal program and select the KitProg3 COM port. Set the serial port parameters to 8N1 and 115200 baud.
3. Program the board using one of the following:
    - Select the code example project in the Project Explorer.
    - In the **Quick Panel**, scroll down, and click **[Project Name] Program (KitProg3_MiniProg4)**.
4. After programming, the code example starts automatically. Confirm that the messages are displayed on the UART terminal:

   *Figure 4. Terminal output on program startup*<BR><img src="./images/terminal-output_menu.png" width="640"/>
    
**Expected behavior**

For demonstration purposes, the following actions are done. The ADC conversion results may vary on different hardware. 
- Set the potentiometer to the lower (clockwise) limit to connect a *0V* voltage level to the ADC input.
- Press the "0" key to start an ADC conversion which should result in "0".
- Press the "2" key to decalibrate the ADC with an offset equal to the maximum offset value.
- Press the "0" key to start an ADC conversion which should result in a value > "0" ("27" in this example)
- Press the "4" key to decalibrate the ADC with a gain equal to the maximum gain value. 
- Press the "0" key to start an ADC conversion which should result in a value > "0" ("36" in this example)
- Press the "1" key to execute the ADC calibration algorithm. The ideal values for offset and gain will be printed on the terminal
- Press the "0" key to start an ADC conversion which should result in "0".
- Set the potentiometer to the upper (counter-clockwise) limit to connect a 5V voltage level to the ADC input.
- Press the "0" key to start an ADC conversion which should result in "4095". <BR><BR>
*Figure 5. Terminal output for the sequence described above*<BR><img src="./images/terminal-output.png" width="640"/>

## References  

Relevant Application notes are:
- AN235305 - GETTING STARTED WITH TRAVEO™ T2G FAMILY MCUS IN MODUSTOOLBOX™
- [AN219755](https://www.infineon.com/dgdl/?fileId=8ac78c8c7cdc391c017d0d3aaebf676a) - USING A SAR ADC IN TRAVEO™ T2G AUTOMOTIVE MICROCONTROLLERS

ModusToolbox™ is available online:
- <https://www.infineon.com/modustoolbox>

Associated TRAVEO™ T2G MCUs can be found on:
- <https://www.infineon.com/cms/en/product/microcontroller/32-bit-traveo-t2g-arm-cortex-microcontroller/>

More code examples can be found on the GIT repository:
- [TRAVEO™ T2G Code examples](https://github.com/orgs/Infineon/repositories?q=mtb-t2g-&type=all&language=&sort=)

For additional training, visit our webpage:  
- [TRAVEO™ T2G training](https://www.infineon.com/cms/en/product/microcontroller/32-bit-traveo-t2g-arm-cortex-microcontroller/32-bit-traveo-t2g-arm-cortex-for-body/traveo-t2g-cyt4bf-series/#!trainings)

For questions and support, use the TRAVEO™ T2G Forum:  
- <https://community.infineon.com/t5/TRAVEO-T2G/bd-p/TraveoII>  


changelog.md
## Changelog  
V1.0.0 2023-08-01 Initial version


main.c
/**********************************************************************************************************************
 * \file main.c
 * \copyright Copyright (C) Infineon Technologies AG 2019
 *
 * Use of this file is subject to the terms of use agreed between (i) you or the company in which ordinary course of
 * business you are acting and (ii) Infineon Technologies AG or its licensees. If and as long as no such terms of use
 * are agreed, use of this file is subject to following:
 *
 * Boost Software License - Version 1.0 - August 17th, 2003
 *
 * Permission is hereby granted, free of charge, to any person or organization obtaining a copy of the software and
 * accompanying documentation covered by this license (the "Software") to use, reproduce, display, distribute, execute,
 * and transmit the Software, and to prepare derivative works of the Software, and to permit third-parties to whom the
 * Software is furnished to do so, all subject to the following:
 *
 * The copyright notices in the Software and this entire statement, including the above license grant, this restriction
 * and the following disclaimer, must be included in all copies of the Software, in whole or in part, and all
 * derivative works of the Software, unless such copies or derivative works are solely in the form of
 * machine-executable object code generated by a source language processor.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 *********************************************************************************************************************/
/*********************************************************************************************************************/
/*-----------------------------------------------------Includes------------------------------------------------------*/
/*********************************************************************************************************************/
#include "cyhal.h"
#include "cybsp.h"
#include "cy_retarget_io.h"

/*********************************************************************************************************************/
/*------------------------------------------------------Macros-------------------------------------------------------*/
/*********************************************************************************************************************/
/* Max and min values for ADC offset and gain */
#define OFFSET_MAX   127
#define OFFSET_MIN  -128
#define GAIN_MAX     15
#define GAIN_MIN    -15

/* Max ADC conversion result */
#define ADC_RESULT_MAX  4095

/*********************************************************************************************************************/
/*-------------------------------------------------Global variables--------------------------------------------------*/
/*********************************************************************************************************************/
/* Calibration setting */
cy_stc_sar2_analog_calibration_conifg_t g_calibrationConfig =
{
    .offset = OFFSET_MAX,
    .gain = 0ul,
};

/*********************************************************************************************************************/
/*---------------------------------------------Function Implementations----------------------------------------------*/
/*********************************************************************************************************************/
/**********************************************************************************************************************
 * Function Name: getAdcValue
 * Summary:
 *  Issues software trigger for a defined channel, waits for ADC to complete its conversion
 *  and returns converted value.
 * Parameters:
 *  channel - channel used for A/D conversion
 * Return:
 *  uint16_t - Measurement value
 **********************************************************************************************************************
 */
static uint16_t getAdcValue(uint32_t channel)
{
    /* Trigger conversion. */
    Cy_SAR2_Channel_SoftwareTrigger(SARADC_HW, channel);
    
    /* Wait for completion. */
    while (CY_SAR2_INT_GRP_DONE != Cy_SAR2_Channel_GetInterruptStatus(SARADC_HW, channel))
    {
    }
    Cy_SAR2_Channel_ClearInterrupt(SARADC_HW, channel, CY_SAR2_INT_GRP_DONE);
    
    /* Get and return the conversion result */
    return (Cy_SAR2_Channel_GetResult(SARADC_HW, channel, NULL));
}

/**********************************************************************************************************************
 * Function Name: printMenu
 * Summary:
 *  Displays a menu on a terminal window to select some actions 
 * Parameters:
 *  none
 * Return:
 *  none
 **********************************************************************************************************************
 */
static void printMenu(void)
{
    /* \x1b[2J\x1b[;H - ANSI ESC sequence for clear screen */
    printf("\x1b[2J\x1b[;H");

    printf("***********************************************************\r\n");
    printf("SAR ADC Calibration\r\n");
    printf("***********************************************************\r\n");
    printf(">> Select one of the options as listed below \r\n");
    printf(">> 0 : Trigger ADC Measurement\r\n");
    printf(">> 1 : Execute ADC Calibration\r\n");
    printf(">> 2 : Set ADC Offset Compensation to OFFSET_MAX\r\n");
    printf(">> 3 : Set ADC Offset Compensation to OFFSET_MIN\r\n");
    printf(">> 4 : Set ADC Gain Compensation to GAIN_MAX\r\n");
    printf(">> 5 : Set ADC Gain Compensation to GAIN_MIN\r\n\r\n");
    printf(">> Others : Print menu\r\n\r\n");
}

/**********************************************************************************************************************
 * Function Name: calibrateOffset
 * Summary:
 *  Performs ADC offset calibration.
 * Parameters:
 *  none
 * Return:
 *  bool - true if the calibration is successful.
 **********************************************************************************************************************
 */
static bool calibrateOffset(void)
{
    uint16_t result;
    int16_t offsetHigh;
    int16_t offsetLow;
    int16_t finalOffsetValue;
    
    /* Set VREF_L as source for the channel and initialize the channel. */
    SARADC_AdcChannel1_config.pinAddress = CY_SAR2_PIN_ADDRESS_VREF_L;
    Cy_SAR2_Channel_Init(SARADC_HW, SARADC_AdcChannel1_IDX, &SARADC_AdcChannel1_config);

    /* Gain configuration set to '0' while offset calibration is executed. */
    g_calibrationConfig.gain = 0;

    /* Go through all offset compensation values from max to min. */
    for (offsetLow = OFFSET_MAX; offsetLow >= OFFSET_MIN; offsetLow -= 1)
    {
        /* Set current offset value in alternate offset register. */
        g_calibrationConfig.offset = (int8_t) offsetLow;
        Cy_SAR2_SetAltAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
        
        /* Get ADC conversion value. */
        result = getAdcValue(SARADC_AdcChannel1_IDX);
        
        /* Offset calibration completed when the ADC conversion value equals to "0". */
        if (result == 0)
        {
            break;
        }
        
        /* Error condition if offset has reached lower bound. */
        if (offsetLow == OFFSET_MIN)
        {
            /* Offset adjustment failed - report error. */
            printf("Error: ADC conversion value > '0' when offset is set to OFFSET_MIN\r\n");
            return false;
        }
    }

    /* Set VREF_H as source for the channel and initialize the channel. */
    SARADC_AdcChannel1_config.pinAddress = CY_SAR2_PIN_ADDRESS_VREF_H;
    Cy_SAR2_Channel_Init(SARADC_HW, SARADC_AdcChannel1_IDX, &SARADC_AdcChannel1_config);
    
    /* Adjust the offset compensation to the smallest offset value possible. */
    g_calibrationConfig.offset = OFFSET_MIN;
    Cy_SAR2_SetAltAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
    
    /* Start ADC conversion and get the A/D value. */
    result = getAdcValue(SARADC_AdcChannel1_IDX);
    
    /* ADC conversion value must be lower than the maximal ADC result value. */
    if (result < ADC_RESULT_MAX)
    {
        /* Go through all offset compensation values from min to max. */
        for (offsetHigh = OFFSET_MIN; offsetHigh <= OFFSET_MAX; offsetHigh += 1)
        {
            /* Set current offset value in alternate offset register. */
            g_calibrationConfig.offset = (int8_t) offsetHigh;
            Cy_SAR2_SetAltAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
            
            /* Get ADC conversion value. */
            result = getAdcValue(SARADC_AdcChannel1_IDX);
            
            /* Offset calibration completed when the ADC conversion value equals to the maximum possible ADC result value. */
            if (result == ADC_RESULT_MAX)
            {
                break;
            }
            
            /* Error condition if offset has reached upper bound. */
            if (offsetHigh == OFFSET_MAX)
            {
                /* Offset adjustment failed - report error. */
                printf("Error: ADC conversion value remains below the maximum possible ADC conversion value when the offset is set OFFSET_MAX.\r\n");
                return false;
            }
        }
    }
    else
    {
        /* Offset adjustment failed - report error. */
        printf("Error: Initial ADC conversion result equals to the maximum possible ADC conversion value when the offset is set to OFFSET_MIN.\r\n");
        return false;
    }
    
    /* Calculate the final offset value and check its expected range */
    finalOffsetValue = ((offsetHigh + offsetLow) / 2) + 2;
    if (finalOffsetValue > OFFSET_MAX)
    {
        /* Offset adjustment failed - report error. */
        printf("Error: Final offset value must not greater than OFFSET_MAX\r\n");
        return false;
    }
    
    /* Update the offset compensation with the calculated value */
    g_calibrationConfig.offset = (int8_t) finalOffsetValue;
    
    /* Print the final calibrated offset value. */
    printf("ADC offset compensation value: %d\r\n", finalOffsetValue);
    return true;
}

/**********************************************************************************************************************
 * Function Name: calibrateGain
 * Summary:
 *  Performs ADC gain calibration
 * Parameters:
 *  none
 * Return:
 *  bool - true if the calibration is successful.
 **********************************************************************************************************************
 */
static bool calibrateGain(void)
{
    int8_t gain = 0;
    uint16_t result;
    
    /* Set gain correction to maximum possible gain value. */
    g_calibrationConfig.gain = GAIN_MAX;
    
    /* Set VREFL voltage to input and init the channel. */
    SARADC_AdcChannel1_config.pinAddress = CY_SAR2_PIN_ADDRESS_VREF_L;
    Cy_SAR2_Channel_Init(SARADC_HW, SARADC_AdcChannel1_IDX, &SARADC_AdcChannel1_config);
    
    /* Apply the initial calibration settings. */
    Cy_SAR2_SetAltAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
    
    /* Start ADC conversion and get the A/D value. */
    result = getAdcValue(SARADC_AdcChannel1_IDX);
    
    /* The ADC conversion result should be higher than the minimum ADC conversion value. */
    if (result > 0)
    {
         /* Decrement the value of gain compensation by "1". */
        for (gain = GAIN_MAX; gain >= GAIN_MIN; gain -= 1)
        {
            /* Set gain to calibration register */
            g_calibrationConfig.gain = gain;
            Cy_SAR2_SetAltAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);

            /* Get ADC conversion value. */
            result = getAdcValue(SARADC_AdcChannel1_IDX);

            /* Gain calibration completed when the ADC conversion value equals to the minimum ADC conversion value. */
            if (result == 0)
            {
                break;
            }
            if (gain == GAIN_MIN)
            {
                /* Gain adjustment failed - report error. */
                printf("Error: ADC conversion value > '0' when the gain is set GAIN_MIN.\r\n");
                return false;
            }
        }
    }
    else
    {
        /* Gain adjustment failed - report error. */
        printf("Error: ADC conversion result equals to '0' when the gain is set GAIN_MAX.\r\n");
        return false;
    }
    
    /* Print calibrated gain value. */
    printf("ADC gain compensation value: %d\r\n", gain);
    return true;
}

/**********************************************************************************************************************
 * Function Name: main
 * Summary:
 *  This is the main function. It configures two ADC channels. One is connected to a potentiometer. 
 *  The second channel is used for ADC calibration. A menu is displayed on a terminal (115200 baud).
 *  The user can choose between different actions: 
 *   - Trigger one ADC conversion of the channel connected to the potentiometer
 *   - De-calibrate ADC by changing offset (OFFSET_MAX, OFFSET_MIN) and gain (GAIN_MAX, GAIN_MIN)
 *   - Execution of ADC calibration algorithm for offset and gain
 *  Status information are displayed for each action. Configurations for ADC and port pins are defined 
 *  in "cycfg_peripherals.h". Any changes should be done in the device configurator. 
 * Parameters:
 *  none
 * Return:
 *  int
 **********************************************************************************************************************
 */
int main(void)
{
    uint16_t adcValue; 
    uint8_t uartReadValue;    

    #if defined (CY_DEVICE_SECURE)
    cyhal_wdt_t wdt_obj;

    /* Clear watchdog timer so that it doesn't trigger a reset */
    result = cyhal_wdt_init(&wdt_obj, cyhal_wdt_get_max_timeout_ms());
    CY_ASSERT(CY_RSLT_SUCCESS == result);
    cyhal_wdt_free(&wdt_obj);
    #endif /* #if defined (CY_DEVICE_SECURE) */

    /* Initialize the device and board peripherals */
    if (cybsp_init()!= CY_RSLT_SUCCESS)
    {
        CY_ASSERT(0);
    }

    /* Enable global interrupts */
    __enable_irq();

    /* Initialize retarget-io to use the debug UART port */
    if (cy_retarget_io_init(CYBSP_DEBUG_UART_TX, CYBSP_DEBUG_UART_RX, CY_RETARGET_IO_BAUDRATE) != CY_RSLT_SUCCESS)
    {
        CY_ASSERT(0);
    }

    /* Print the menu to list all options */
    printMenu();
    
    /* Set ePASS MMIO reference buffer mode for bandgap voltage */
    Cy_SAR2_SetReferenceBufferMode(PASS0_EPASS_MMIO, CY_SAR2_REF_BUF_MODE_ON);

    /* Enable the Diagnostic Reference */
    Cy_SAR2_Diag_Enable(SARADC_HW);
    
     /* Initialize the ADC */
    Cy_SAR2_Init(SARADC_HW, &SARADC_config);
    
    /* Enable ADC */
    Cy_SAR2_Enable(SARADC_HW);    
    
    for (;;)
    {
        /* Check if key was pressed and perform actions */
        if (cyhal_uart_getc(&cy_retarget_io_uart_obj, &uartReadValue, 1) == CY_RSLT_SUCCESS)
        {
            switch (uartReadValue)
            {
                /* Trigger a single ADC conversion of potentiometer voltage and display it on terminal */
                case '0':
                    printf(">> 0 : Trigger ADC Measurement\r\n");
                    adcValue = getAdcValue(SARADC_AdcChannel0_IDX);
                    printf("ADC-Value: %d\r\n", adcValue);  
                    break;
                
                /* Execute ADC calibration algorithm for offset and gain */
                case '1':
                    printf(">> 1 : Execute ADC Calibration\r\n");
                    if(calibrateOffset() == true)
                    {
                        printf("Offset Calibration successful\r\n");  
                    }
                    else
                    {
                        printf("Offset Calibration error\r\n");  
                    }

                    if(calibrateGain() == true)
                    {
                        printf("Gain Calibration successful\r\n");  
                    }
                    else
                    {
                        printf("Gain Calibration error\r\n");  
                    }
                    
                    /* Update regular calibration values based on calibration data
                     * found in alternate calibration register. */
                    Cy_SAR2_TriggerCalibrationUpdate(SARADC_HW);
                    break;
                
                /* De-calibrate ADC with offset set to OFFSET_MAX */
                case '2':
                    printf(">> 2 : Set ADC Offset Compensation to OFFSET_MAX\r\n");
                    
                    /* Set regular calibration offset to OFFSET_MAX to de-calibrate the ADC. */
                    g_calibrationConfig.offset = OFFSET_MAX;
                    Cy_SAR2_SetAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
                    printf("ADC Offset set to OFFSET_MAX\r\n");  
                    break;
                
                /* De-calibrate ADC with offset set to OFFSET_MIN */
                case '3':
                    printf(">> 3 : Set ADC Offset Compensation to OFFSET_MIN\r\n");

                    /* Set regular calibration offset to OFFSET_MIN to de-calibrate the ADC. */
                    g_calibrationConfig.offset = OFFSET_MIN;
                    Cy_SAR2_SetAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
                    printf("ADC Offset set to OFFSET_MIN\r\n");  
                    break;
                
                /* De-calibrate ADC with gain set to GAIN_MAX */
                case '4':
                    printf(">> 4 : Set ADC Gain Compensation to GAIN_MAX\r\n");
                    
                    /* Set regular calibration gain to GAIN_MAX to de-calibrate the ADC. */
                    g_calibrationConfig.gain = GAIN_MAX;
                    Cy_SAR2_SetAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
                    printf("ADC Gain set to GAIN_MAX\r\n");  
                    break;
                
                /* De-calibrate ADC with offset set to GAIN_MIN */
                case '5':
                    printf(">> 5 : Set ADC Gain Compensation to GAIN_MIN\r\n\r\n");
                    
                    /* Set regular calibration gain to GAIN_MIN to de-calibrate the ADC. */
                    g_calibrationConfig.gain = GAIN_MIN;
                    Cy_SAR2_SetAnalogCalibrationValue(SARADC_HW, &g_calibrationConfig);
                    printf("ADC Gain set to GAIN_MIN\r\n");  
                    break;
                
                /* Print menu for all other keys pressed */
                default:
                    printMenu();
                    break;                    
            }
        }        
    }
}

/* [] END OF FILE */
